### PR TITLE
Adds source and sink filtering ability to sapp

### PR DESCRIPTION
**Pull request by abishekvashok**
Adds ability to filter for issues in sapp with SourceName, SourceKind,
SinkName, and SinkKind options.

Tweaks the React UI to display the options in the FilterForm and
modifies the query that it makes with the backend flask server.

Implements the ability on the backend server to filter for issues based
on the given SinkName, SinkKind, SourceName, and SourceKind
filters/options.

Filtering by SourceName, SourceKind, SinkName, and SinkKind are via
additional graphQL subqueries that are concatenated and grouped just
like how features are filtered.

Modifies the HasAny IssuePredicate to have a single class for features
and the new mentioned options

Modifies tests to confront with the new filtering options.

Signed-off-by: Abishek V Ashok <abishekvashok@gmail.com>
Fixes: https://github.com/MLH-Fellowship/pyre-check/issues/14

### CODE REVIEW

**Comment by abishekvashok** [comment]
Significant attention has been given to fix tests that were broken by these changes. And I've fixed those tests that were breaking due to these changes.

Working Screenshots:

Screenshot 2021-07-09 at 12 38 56 PM
Screenshot 2021-07-09 at 12 38 49 PM
Screenshot 2021-07-09 at 12 38 38 PM

**Review by r0rshark — CHANGES_REQUESTED** [review]
This looks very good, thanks for sharing the pictures as well :)
UI filter section looks like it's starting to get pretty large , do you think we can instead have two sections (collapsible would be awesome) one for sources and one for sinks, each with a "name" and a "kind" input.
Moreover I see some tests failing https://github.com/facebook/sapp/pull/23/checks?check_run_id=3033095757 can we address them :)
Also pyre is showing a couple of type errors in the added code (https://github.com/facebook/sapp/runs/3033104881?check_suite_focus=true):
```
sapp/ui/filter_predicates.py:118:12 Incompatible return type [7]: Expected `List[IssueQueryResult]` but got `List[List[IssueQueryResult]]`.
sapp/ui/schema.py:183:12 Incompatible parameter type [6]: Expected `List[str]` for 4th positional only parameter to call `Filter.from_query` but got `Optional[List[str]]`.
```

**Comment by abishekvashok** [comment]
> UI filter section looks like it's starting to get pretty large , do you think we can instead have two sections (collapsible would be awesome) one for sources and one for sinks, each with a "name" and a "kind" input.

For sure!

I will implement this and fix tests to get back.

**Comment by abishekvashok** [comment]
Hey @r0rshark had to use pyre-ignore to fix pyre-check but fixed the tests, and implemented the collapsible

Screenshot 2021-07-12 at 7 57 06 PM

**Comment by dkgi** [comment]
Oh nice, any chance we could fix the alignment issue in the UI? I.e. make sure the text inputs line up with the rest of the form?

It'd also be nice if this was visually a bit less heavy. I like the "ghost" version with the transparent background from the docs (https://ant.design/components/collapse/).

**Comment by abishekvashok** [comment]
Hey @dkgi thanks for the nice suggestions.
I have implemented the changes:
Screenshot 2021-07-13 at 10 35 57 AM

I guess it looks more uniform and minimal now :D

File: sapp/ui/frontend/src/Filter.css [hunk 1]
```diff
@@ -21,3 +21,16 @@
...
+}
+
+.ant-collapse-header {
+    padding: 12px 16px 12px 20px !important;
+}
+
+.ant-collapse-content-box{
+    padding: 1px !important;
+}
+
```
Comment (r0rshark) [inline sapp/ui/frontend/src/Filter.css:36]: @abishekvashok  Internal formatter complained about this line. Can we delete it please :)

**Comment by abishekvashok** [comment]
@r0rshark done :)

**Comment by dkgi** [comment]
That looks very nice! Thank you :)

**Pull request closed without merging**